### PR TITLE
Issue 1402

### LINKED ISSUE #1402: WPS432 should preserve the literal representation
# Bug report

This code
```python
import pathlib

path = pathlib.Path('.')
path.chmod(0o700)
```
causes
```console
WPS432: Found magic number: 448
```

It's unobvious that it's the same number and it's impossible to use `448` in grep because it won't find anything.

## What's wrong

Non-decimal numbers are reported as decimal. Also, maybe for things like `chmod()` it's okay to allow such "magic numbers" because this form directly relates to a normal syscall or a CLI command one would make.

## How is that should be

The formatting should be preserved. The error should've been `WPS432: Found magic number: 0o700` instead.

## System information

N/A

### PR DESCRIPTION
**Pull request by mihsan96**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] I have created at least one test case for the changes I have made
- [ ] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: wemake_python_styleguide/visitors/base.py [hunk 1]
```diff
@@ -286,3 +286,58 @@ def run(self) -> None:
...
+        options: ValidatedOptions,
+        tree: ast.AST,
+        file_tokens: Sequence[tokenize.TokenInfo],
+        **kwargs,
+    ) -> None:
+        """Creates new ``ast`` based instance with tokens."""
+        super().__init__(options, **kwargs)
+        self.tree = tree
+        self.file_tokens = file_tokens
+        self.token_index = -1
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:305]:
```suggestion
        self._token_index = -1
        self._token_dict: dict[tuple[int, int], tokenize.TokenInfo] = {}
```

File: wemake_python_styleguide/visitors/base.py [hunk 2]
```diff
@@ -286,3 +286,58 @@ def run(self) -> None:
...
+
+    @final
+    @classmethod
+    def from_checker(cls: type['BaseNodeTokenVisitor'], checker):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:309]:
```suggestion
    def from_checker(cls: type['BaseNodeTokenVisitor'], checker) -> Self:
```

File: wemake_python_styleguide/visitors/base.py [hunk 3]
```diff
@@ -286,3 +286,58 @@ def run(self) -> None:
...
+    @final
+    def run(self) -> None:
+        """Recursively visits all ``ast`` nodes and create token_dict."""
+        self._create_token_dict()
+        self.visit(self.tree)
+        self._post_visit()
+
+    def _create_token_dict(self) -> None:
+        """Create a token dict."""
+        self.token_dict: dict[tuple[int, int], tokenize.TokenInfo] = {}
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:341]:
```suggestion
```

**Review by sobolevn — COMMENTED** [review]
Thanks! This is a really cool way to solve this!

**Comment by mihsan96** [comment]
My bad, I didn't notice these tests. Is replace underscore good way to fix this?

File: wemake_python_styleguide/visitors/ast/builtins.py [hunk 4]
```diff
@@ -197,9 +197,9 @@ def _check_is_magic(self, node: ast.Constant) -> None:
 
         if isinstance(node.value, int) and node.value <= self._non_magic_modulo:
             return
-
+        real_value = self._token_dict[node.lineno, node.col_offset].string
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/builtins.py:200]: This can also be extracted to some helper method.

```suggestion
        try:
            token = self._token_dict[node.lineno, node.col_offset]
        except KeyError:
            # For some reason, the token was not found
            real_value = str(node.value)
        else:
            real_value = token.string
```

File: wemake_python_styleguide/visitors/base.py [hunk 5]
```diff
@@ -286,3 +286,60 @@ def run(self) -> None:
...
+    def run(self) -> None:
+        """Recursively visits all ``ast`` nodes and create token_dict."""
+        self._create_token_dict()
+        self.visit(self.tree)
+        self._post_visit()
+
+    def _create_token_dict(self) -> None:
+        """Create a token dict."""
+        for token in self.file_tokens:
+            self._token_dict[token.start] = token
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:345]:
```suggestion
        self._token_dict = {token.start: token for token in self.file_tokens}
```

File: wemake_python_styleguide/visitors/base.py [hunk 6]
```diff
@@ -286,3 +286,60 @@ def run(self) -> None:
...
+        """
+        Visits a node.
+
+        Modified version of :class:`ast.NodeVisitor.visit` method.
+        We need this to modify how visitors route.
+
+        Why? Because python3.8 now uses ``visit_Constant`` instead of old
+        methods like ``visit_Num``, ``visit_Str``, ``visit_Bytes``, etc.
+
+        Some classes do redefine this method to catch all nodes. This is valid.
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:331]: Can you please instead link to the `BaseNodeVisitor` docs? We don't need to duplicate the reasoning here.

File: wemake_python_styleguide/visitors/base.py [hunk 7]
```diff
@@ -286,3 +286,60 @@ def run(self) -> None:
...
+        """
+        return route_visit(self, tree)
+
+    @final
+    def run(self) -> None:
+        """Recursively visits all ``ast`` nodes and create token_dict."""
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/base.py:337]:
```suggestion
        """Recursively visits all ``ast`` nodes and create ``token_dict``."""
```

File: tests/test_visitors/test_ast/test_builtins/test_numbers/test_approximate_constants.py [hunk 8]
```diff
@@ -24,10 +26,15 @@ def test_violation_on_approximate_constants(
...
 ):
     """Ensures that usage of approximate constants not allowed."""
-    tree = parse_ast_tree(f'my_const = {variable_value}')
-    visitor = WrongNumberVisitor(default_options, tree=tree)
+    variable_str = f'my_const = {variable_value}'
+    tree = parse_ast_tree(variable_str)
+    file_tokens = parse_tokens(variable_str)
+    visitor = WrongNumberVisitor(
+        default_options, tree=tree, file_tokens=file_tokens
+    )
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_builtins/test_numbers/test_approximate_constants.py:37]: Please, add some extra `variable_value` examples here: we need to test that the formatting is preserved.

For example: `3.1_415` will be formatted exactly like `3.1_415` and not like `3.1415`

File: tests/test_visitors/test_ast/test_builtins/test_numbers/test_magic_numbers.py [hunk 9]
```diff
@@ -200,11 +210,16 @@ def test_magic_number_warning(
...
+    parse_tokens,
 ):
     """Testing that magic numbers in this code are warnings."""
-    tree = parse_ast_tree(mode(code.format(number)))
+    formated_code = mode(code.format(number))
+    tree = parse_ast_tree(formated_code)
+    file_tokens = parse_tokens(formated_code)
 
-    visitor = WrongNumberVisitor(default_options, tree=tree)
+    visitor = WrongNumberVisitor(
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_builtins/test_numbers/test_magic_numbers.py:220]: Please, add octal, hex, and binary numbers to the test examples here as well.

**Review by sobolevn — COMMENTED** [review]
Thank you, almost done! 
Btw, happy new year! 🌲 🎅

**Comment by sobolevn** [comment]
Oh wow, looks like paths are specific to your OS - Windows. I think that this should be fixed in https://github.com/wemake-services/wemake-python-styleguide/blob/92fa698cf3edabfe94bffafaa002c9f5f31f86df/tests/test_formatter/test_formatter_output.py#L31-L41

**Comment by mihsan96** [comment]
> Oh wow, looks like paths are specific to your OS - Windows. I think that this should be fixed in
> 
> https://github.com/wemake-services/wemake-python-styleguide/blob/92fa698cf3edabfe94bffafaa002c9f5f31f86df/tests/test_formatter/test_formatter_output.py#L31-L41

I feel I don't need to fix this in this pr. Probably some where else is also same problems.

File: tests/test_formatter/__snapshots__/test_formatter_output.ambr [hunk 10]
```diff
@@ -957,9 +957,10 @@
   '''
   
   [4m[1mtests/fixtures/notebook.ipynb[0m[0m
+    0:1      WPS453 Found executable mismatch: file is executable but no shebang is present
```
Comment (sobolevn) [inline tests/test_formatter/__snapshots__/test_formatter_output.ambr:960]: this does not look related.

Comment (mihsan96) [reply tests/test_formatter/__snapshots__/test_formatter_output.ambr:960]: have no idea how it happens. i`m just run 
poetry run pytest --snapshot-update
and all test is passed

Comment (mihsan96) [reply tests/test_formatter/__snapshots__/test_formatter_output.ambr:960]: is it bad idea to fix this manually?

Comment (sobolevn) [reply tests/test_formatter/__snapshots__/test_formatter_output.ambr:960]: it is fine, just remove the line

Comment (sobolevn) [inline tests/test_formatter/__snapshots__/test_formatter_output.ambr:960]:
```suggestion
```

**Review by sobolevn — APPROVED** [review]
Thank you! Happy New Year!

File: tests/test_formatter/__snapshots__/test_formatter_output.ambr (hunk 10 above)
Comment (mihsan96) [reply tests/test_formatter/__snapshots__/test_formatter_output.ambr]: i think this happens because of wsl. but have no idea why

**Comment by mihsan96** [comment]
> Thank you! Happy New Year!

Thank you for your help in implementing this PR! Happy New Year!
